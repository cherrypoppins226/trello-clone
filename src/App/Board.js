import React from "react";
import { withStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Modal from "@material-ui/core/Modal";
import PropTypes from "prop-types";
import Popover from "@material-ui/core/Popover";
import CardsList from "./CardsList";
import QuickEditCard from "./QuickEditCard";
import ActionsMenu from "./CardsList/ActionsMenu";
import EditCard from "./EditCard";
import { cardDescription } from "./CardsList/Card";

const styles = {
  root: {
    padding: 8,
    margin: 0,
    height: "100%",
    width: "100%"
  },
  cardListContainer: {
    width: 280
  },
  cardsListActionsPopover: {
    backgroundColor: "transparent"
  }
};

class Board extends React.Component {
  state = {
    listBeingEdited: null,
    cardBeingQuickEdited: null,
    cardBeingEdited: null
  };

  render() {
    const { classes, lists } = this.props;
    const cardTitle = (v => (v ? cardDescription(v) : ""))(
      this.state.cardBeingEdited || this.state.cardBeingQuickEdited
    );
    return (
      <>
        <Grid container spacing={16} wrap="nowrap" className={classes.root}>
          {Object.entries(lists).map(([k, v]) => {
            return (
              <Grid item key={k} className={classes.cardListContainer}>
                <CardsList
                  title={k}
                  cards={v}
                  onQuickEditCard={card =>
                    this.setState({ cardBeingQuickEdited: card })
                  }
                  onEditCard={card => this.setState({ cardBeingEdited: card })}
                  onEditList={list => this.setState({ listBeingEdited: list })}
                />
              </Grid>
            );
          })}
        </Grid>
        <Popover
          anchorEl={this.state.listBeingEdited}
          container={this}
          open={Boolean(this.state.listBeingEdited)}
          onClose={_ => this.setState({ listBeingEdited: null })}
          TransitionProps={{ timeout: 0 }}
          elevation={1}
          anchorOrigin={{ vertical: "bottom", horizontal: "left" }}
          transformOrigin={{ vertical: "top", horizontal: "left" }}
          PaperProps={{ classes: { root: classes.cardsListActionsPopover } }}
        >
          <ActionsMenu
            onMenuItemClick={_ => this.setState({ listBeingEdited: null })}
          />
        </Popover>
        <Modal
          container={this}
          open={Boolean(this.state.cardBeingQuickEdited)}
          onClose={_ => this.setState({ cardBeingQuickEdited: null })}
          BackdropProps={{ id: "quickEditCardBackdrop" }}
        >
          <QuickEditCard
            title={cardTitle}
            anchorEl={this.state.cardBeingQuickEdited}
          />
        </Modal>
        <Modal
          container={this}
          open={Boolean(this.state.cardBeingEdited)}
          onClose={_ => this.setState({ cardBeingEdited: null })}
          BackdropProps={{ id: "editCardBackdrop" }}
        >
          <EditCard title={cardTitle} />
        </Modal>
      </>
    );
  }
}

const View = withStyles(styles)(Board);

View.propTypes = {
  lists: PropTypes.object.isRequired
};

export default View;
