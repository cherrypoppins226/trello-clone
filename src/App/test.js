import { getByText } from "dom-testing-library";
import { Simulate } from "react-testing-library";

import { render } from "./utils/dom";
import { labelledBy, describedBy } from "./utils";
import proxies from "../cosmos.proxies";
import * as labels from "./labels";
import * as fixtures from "./fixtures";

describe("cards list", () => {
  it("opens list actions menu", async () => {
    const { container } = await render(fixtures.App.default, proxies);
    const getMenu = () =>
      container.querySelector(
        describedBy(labels.cardsListActionsMenuDescription.id)
      );
    expect(getMenu()).toBeNull();
    Simulate.click(
      container.querySelector(labelledBy(labels.cardsListActionsMenu.id))
    );
    expect(getMenu()).not.toBeNull();
  });

  it("adds a card", async () => {
    const { container } = await render(
      fixtures.CardsList.CardsList.default,
      proxies
    );
    const liveList = container.querySelector("ul");
    const countBefore = liveList.childElementCount;
    const lastBefore = liveList.lastElementChild;
    Simulate.click(getByText(container, "Add a card..."));
    expect(liveList.childElementCount).toBe(countBefore + 1);
    expect(liveList.lastElementChild).not.toBe(lastBefore);
    expect(liveList.lastElementChild.tagName).toBe(lastBefore.tagName);
  });
});

describe("cards list card", () => {
  const testModal = async (editButtonSelector, modalSelector, textSelector) => {
    const { container } = await render(fixtures.App.default, proxies);
    const card = container.querySelector("[data-cardid='1']");
    expect(container.querySelector(modalSelector)).toBeNull();
    Simulate.click(
      editButtonSelector ? card.querySelector(editButtonSelector) : card
    );
    expect(container.querySelector(modalSelector)).not.toBeNull();
    const cardDescription = node =>
      node.querySelector(labelledBy(labels.card.id)).textContent;
    expect(
      container.querySelector(modalSelector).querySelector(textSelector)
        .textContent
    ).toBe(cardDescription(card));
  };

  it("opens quick edit card modal", async () => {
    await testModal(
      labelledBy(labels.quickEditCard.id),
      describedBy(labels.quickEditCardDescription.id),
      "textarea"
    );
  });

  it("opens edit card modal", async () => {
    await testModal(
      null,
      describedBy(labels.editCardDescription.id),
      "[role='heading']"
    );
  });
});
