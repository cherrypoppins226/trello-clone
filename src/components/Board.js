import React from "react";
import { withStyles } from "material-ui/styles";
import Grid from "material-ui/Grid";
import PropTypes from "prop-types";
import Popover from "material-ui/Popover";
import CardsList from "./CardsList";
import QuickEditCard from "./QuickEditCard";
import ActionsMenu from "./CardsList/ActionsMenu";
import EditCard from "./EditCard";
import * as Labels from "./labels";

const styles = {
  root: {
    padding: 8,
    margin: 0,
    height: "100%",
    width: "100%"
  },
  cardListContainer: {
    width: 280
  }
};

class Board extends React.Component {
  state = {
    listBeingEdited: null,
    cardBeingQuickEdited: null,
    cardBeingEdited: null
  };

  render() {
    const { classes, lists } = this.props;
    return (
      <>
        <Grid container spacing={16} wrap="nowrap" className={classes.root}>
          {Object.entries(lists).map(([k, v]) => {
            return (
              <Grid item key={k} className={classes.cardListContainer}>
                <CardsList
                  title={k}
                  cards={v}
                  onQuickEditCard={card =>
                    this.setState({ cardBeingQuickEdited: card })
                  }
                  onEditCard={card => this.setState({ cardBeingEdited: card })}
                  onEditList={list => this.setState({ listBeingEdited: list })}
                />
              </Grid>
            );
          })}
        </Grid>
        <Popover
          aria-describedby={Labels.cardsListActionsMenuDescription.id}
          anchorEl={this.state.listBeingEdited}
          container={this}
          open={Boolean(this.state.listBeingEdited)}
          onClose={_ => this.setState({ listBeingEdited: null })}
          transitionDuration={0}
          elevation={1}
          anchorOrigin={{ vertical: "bottom", horizontal: "left" }}
          transformOrigin={{ vertical: "top", horizontal: "left" }}
        >
          <ActionsMenu
            onMenuItemClick={_ => this.setState({ listBeingEdited: null })}
          />
        </Popover>
        <QuickEditCard
          container={this}
          card={this.state.cardBeingQuickEdited}
          onClose={_ => this.setState({ cardBeingQuickEdited: null })}
        />
        <EditCard
          container={this}
          card={this.state.cardBeingEdited}
          onClose={_ => this.setState({ cardBeingEdited: null })}
        />
      </>
    );
  }
}

const View = withStyles(styles)(Board);

View.propTypes = {
  lists: PropTypes.object.isRequired
};

export default View;
