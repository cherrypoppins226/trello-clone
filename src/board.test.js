import React from "react";
import faker from "faker";
import { shallow } from "./test/enzymeShallowUtils";
import Board from "./board";

faker.seed(1);

const lists = {
  [faker.lorem.sentence()]: [faker.lorem.sentence(), faker.lorem.sentence()],
  [faker.lorem.sentence()]: [faker.lorem.sentence(), faker.lorem.sentence()]
};

it("initial render", () => {
  const tree = shallow(<Board lists={lists} />);
  expect(tree).toMatchSnapshot();
});
