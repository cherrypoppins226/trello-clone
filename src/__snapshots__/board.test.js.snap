// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`board initial render 1`] = `
<WithStyles(GridWrapper)
  className="View-grid-1"
  container={true}
  spacing={16}
  wrap="nowrap"
>
  <WithStyles(GridWrapper)
    className="View-cardListContainer-2"
    item={true}
    key="Repellat quisquam recusandae alias consequuntur corporis."
  >
    <WithStyles(View)
      cards={
        Array [
          "Ratione ut sunt qui amet iure ut libero qui recusandae.",
          "Nulla quam ipsam nobis cupiditate sed dignissimos.",
        ]
      }
      title="Repellat quisquam recusandae alias consequuntur corporis."
    />
  </WithStyles(GridWrapper)>
  <WithStyles(GridWrapper)
    className="View-cardListContainer-2"
    item={true}
    key="Incidunt accusantium sed libero repudiandae esse blanditiis natus et eos."
  >
    <WithStyles(View)
      cards={
        Array [
          "Velit omnis et porro ut et ipsam explicabo eligendi occaecati.",
          "Et saepe eum dicta eum eaque enim ipsum inventore debitis.",
        ]
      }
      title="Incidunt accusantium sed libero repudiandae esse blanditiis natus et eos."
    />
  </WithStyles(GridWrapper)>
</WithStyles(GridWrapper)>
`;
